diff --git a/theories/dune b/theories/dune
index ae31adb..faa1c1d 100644
--- a/theories/dune
+++ b/theories/dune
@@ -8,27 +8,27 @@
  ; (per_file
  ;  (Init/*.v -> -boot))
  (libraries
-   ltac_plugin
-   tauto_plugin
+   coq-core.plugins.ltac
+   coq-core.plugins.tauto
 
-   cc_plugin
-   firstorder_plugin
+   coq-core.plugins.cc
+   coq-core.plugins.firstorder
 
-   number_string_notation_plugin
+   coq-core.plugins.number_string_notation
 
-   btauto_plugin
-   rtauto_plugin
+   coq-core.plugins.btauto
+   coq-core.plugins.rtauto
 
-   ring_plugin
-   nsatz_plugin
+   coq-core.plugins.ring
+   coq-core.plugins.nsatz
 
-   zify_plugin
-   micromega_plugin
+   coq-core.plugins.zify
+   coq-core.plugins.micromega
 
-   funind_plugin
+   coq-core.plugins.funind
 
-   ssreflect_plugin
-   derive_plugin))
+   coq-core.plugins.ssreflect
+   coq-core.plugins.derive))
 
 (include_subdirs qualified)
 
diff --git a/user-contrib/Ltac2/dune b/user-contrib/Ltac2/dune
index 785cdfe..a5adfa6 100644
--- a/user-contrib/Ltac2/dune
+++ b/user-contrib/Ltac2/dune
@@ -3,4 +3,13 @@
  (package coq-stdlib)
  (synopsis "Ltac2 tactic language")
  (flags -w -deprecated-native-compiler-option)
- (libraries coq-core.plugins.ltac2))
+ (libraries  ; yes?
+   coq-core.plugins.ltac
+   coq-core.plugins.tauto
+
+   coq-core.plugins.cc
+   coq-core.plugins.firstorder
+
+   coq-core.plugins.number_string_notation
+
+   coq-core.plugins.ltac2))
