diff --git a/engine/logic_monad.ml b/engine/logic_monad.ml
index 3c383b2..dd8eb57 100644
--- a/engine/logic_monad.ml
+++ b/engine/logic_monad.ml
@@ -144,6 +144,17 @@ type ('a, 'b, 'e) list_view =
   | Nil of 'e
   | Cons of 'a * ('e -> 'b)
 
+
+type 'a thunk =
+    | Fin of 'a
+    | TailCall of (unit -> 'a thunk)
+
+let rec trampoline r = match r with
+    | Fin a -> a
+    | TailCall f -> trampoline (f ())
+
+
+
 module BackState =
 struct
 
@@ -180,11 +191,11 @@ struct
 
   type ('a, 'i, 'o, 'e) t =
       { iolist : 'r. 'i -> ('e -> 'r NonLogical.t) ->
-                     ('a -> 'o -> ('e -> 'r NonLogical.t) -> 'r NonLogical.t) ->
-                     'r NonLogical.t }
+                     ('a -> 'o -> ('e -> 'r NonLogical.t) -> 'r NonLogical.t thunk) ->
+                     'r NonLogical.t thunk }
 
   let return x =
-    { iolist = fun s nil cons -> cons x s nil }
+    { iolist = fun s nil cons -> TailCall (fun () -> cons x s nil) }
 
   let (>>=) m f =
     { iolist = fun s nil cons ->
@@ -198,16 +209,16 @@ struct
     { iolist = fun s nil cons -> m.iolist s nil (fun x s next -> cons (f x) s next) }
 
   let zero e =
-    { iolist = fun _ nil cons -> nil e }
+    { iolist = fun _ nil cons -> Fin (nil e) }
 
   let plus m1 m2 =
-    { iolist = fun s nil cons -> m1.iolist s (fun e -> (m2 e).iolist s nil cons) cons }
+    { iolist = fun s nil cons -> m1.iolist s (fun e -> trampoline @@ (m2 e).iolist s nil cons) cons }
 
   let ignore m =
     { iolist = fun s nil cons -> m.iolist s nil (fun _ s next -> cons () s next) }
 
   let lift m =
-    { iolist = fun s nil cons -> NonLogical.(m >>= fun x -> cons x s nil) }
+    { iolist = fun s nil cons -> Fin NonLogical.(m >>= fun x -> trampoline @@ cons x s nil) }
 
   (** State related *)
 
@@ -246,30 +257,31 @@ struct
     { iolist = fun s0 nil cons ->
       let next = function
       | Nil e -> nil e
-      | Cons ((x, s), l) -> cons x s (fun e -> (reflect (l e)).iolist s0 nil cons)
+      | Cons ((x, s), l) -> trampoline @@ cons x s (fun e -> trampoline @@ (reflect (l e)).iolist s0 nil cons)
       in
-      NonLogical.(m >>= next)
+      Fin NonLogical.(m >>= next)
     }
 
   let split m : ((_, _, _) list_view, _, _, _) t =
     let rnil e = NonLogical.return (Nil e) in
-    let rcons p s l = NonLogical.return (Cons ((p, s), l)) in
+    let rcons p s l = Fin (NonLogical.return (Cons ((p, s), l))) in
     { iolist = fun s nil cons ->
       let open NonLogical in
-      m.iolist s rnil rcons >>= begin function
-      | Nil e -> cons (Nil e) s nil
+      Fin ((trampoline @@
+      m.iolist s rnil rcons) >>= begin function
+      | Nil e -> trampoline @@ cons (Nil e) s nil
       | Cons ((x, s), l) ->
         let l e = reflect (l e) in
-        cons (Cons (x, l)) s nil
-      end }
+        trampoline @@ cons (Cons (x, l)) s nil
+      end) }
 
   let run m s =
     let rnil e = NonLogical.return (Nil e) in
     let rcons x s l =
       let p = (x, s) in
-      NonLogical.return (Cons (p, l))
+      Fin (NonLogical.return (Cons (p, l)))
     in
-    m.iolist s rnil rcons
+    trampoline @@ m.iolist s rnil rcons
 
   let repr x = x
 end
@@ -349,7 +361,7 @@ struct
       writer) and efficiency. *)
 
   let get =
-    { iolist = fun s nil cons -> cons s.sstate s nil }
+    { iolist = fun s nil cons -> TailCall (fun () -> cons s.sstate s nil) }
 
   let set (sstate : P.s) =
     { iolist = fun s nil cons -> cons () { s with sstate } nil }
@@ -378,8 +390,8 @@ struct
     let rnil e = NonLogical.return (Nil e) in
     let rcons x s l =
       let p = (x, s.sstate, s.wstate, s.ustate) in
-      NonLogical.return (Cons (p, l))
+      Fin (NonLogical.return (Cons (p, l)))
     in
-    m.iolist s rnil rcons
+    trampoline @@ m.iolist s rnil rcons
 
  end
diff --git a/pretyping/constr_matching.ml b/pretyping/constr_matching.ml
index 7d1bb5e..539ca77 100644
--- a/pretyping/constr_matching.ml
+++ b/pretyping/constr_matching.ml
@@ -447,14 +447,25 @@ let matches_head env sigma pat c =
     | _ -> c in
   matches env sigma pat head
 
+
+type 'a thunk =
+    | Fin of 'a
+    | TailCall of (unit -> 'a thunk)
+
+let rec trampoline r = match r with
+    | Fin a -> a
+    | TailCall f -> trampoline (f ())
+
+let tlift f () = trampoline (f ())
+
 (* Tells if it is an authorized occurrence and if the instance is closed *)
 let authorized_occ env sigma closed pat c mk_ctx =
   try
     let subst = matches_core_closed env sigma pat c in
     if closed && Id.Map.exists (fun _ c -> not (closed0 sigma c)) (snd subst)
-    then (fun next -> next ())
-    else (fun next -> mkresult subst (lazy (mk_ctx (mkMeta special_meta))) next)
-  with PatternMatchingFailure -> (fun next -> next ())
+    then (fun next -> TailCall next)
+    else (fun next -> Fin (mkresult subst (lazy (mk_ctx (mkMeta special_meta))) (tlift next)))
+  with PatternMatchingFailure -> (fun next -> TailCall next)
 
 let subargs env v = Array.map_to_list (fun c -> (env, c)) v
 
@@ -469,42 +480,42 @@ let sub_match ?(closed=true) env sigma pat c =
       | [c1] -> mk_ctx (mkCast (c1, k, c2))
       | _ -> assert false
       in
-      try_aux [env, c1] next_mk_ctx next
+      TailCall (fun () -> try_aux [env, c1] next_mk_ctx next)
   | Lambda (x,c1,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkLambda (x, c1, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalAssum (x,c1)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | Prod (x,c1,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkProd (x, c1, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalAssum (x,c1)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | LetIn (x,c1,t,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkLetIn (x, c1, t, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalDef (x,c1,t)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | App (c1,lc) ->
      let lc1 = Array.sub lc 0 (Array.length lc - 1) in
      let app = mkApp (c1,lc1) in
      let mk_ctx = function
        | [app';c] -> mk_ctx (mkApp (app',[|c|]))
        | _ -> assert false in
-     try_aux [(env, app); (env, Array.last lc)] mk_ctx next
+     TailCall (fun () -> try_aux [(env, app); (env, Array.last lc)] mk_ctx next)
   | Case (ci,hd,c1,lc) ->
       let next_mk_ctx = function
       | c1 :: hd :: lc -> mk_ctx (mkCase (ci,hd,c1,Array.of_list lc))
       | _ -> assert false
       in
       let sub = (env, c1) :: (env, hd) :: subargs env lc in
-      try_aux sub next_mk_ctx next
+      TailCall (fun () -> try_aux sub next_mk_ctx next)
   | Fix (indx,(names,types,bodies as recdefs)) ->
     let nb_fix = Array.length types in
     let next_mk_ctx le =
@@ -512,7 +523,7 @@ let sub_match ?(closed=true) env sigma pat c =
       mk_ctx (mkFix (indx,(names, Array.of_list ntypes, Array.of_list nbodies))) in
     let env' = push_rec_types recdefs env in
     let sub = subargs env types @ subargs env' bodies in
-    try_aux sub next_mk_ctx next
+    TailCall (fun () -> try_aux sub next_mk_ctx next)
   | CoFix (i,(names,types,bodies as recdefs)) ->
     let nb_fix = Array.length types in
     let next_mk_ctx le =
@@ -520,15 +531,15 @@ let sub_match ?(closed=true) env sigma pat c =
       mk_ctx (mkCoFix (i,(names, Array.of_list ntypes, Array.of_list nbodies))) in
     let env' = push_rec_types recdefs env in
     let sub = subargs env types @ subargs env' bodies in
-    try_aux sub next_mk_ctx next
+    TailCall (fun () -> try_aux sub next_mk_ctx next)
   | Proj (p,c') ->
     begin try
       let term = Retyping.expand_projection env sigma p c' [] in
-      aux env term mk_ctx next
-    with Retyping.RetypeError _ -> next ()
+      TailCall (fun () -> aux env term mk_ctx next)
+    with Retyping.RetypeError _ -> TailCall next
     end
   | Construct _|Ind _|Evar _|Const _|Rel _|Meta _|Var _|Sort _|Int _|Float _ ->
-    next ()
+    TailCall next
   in
   here next
 
@@ -536,16 +547,16 @@ let sub_match ?(closed=true) env sigma pat c =
   and try_aux lc mk_ctx next =
     let rec try_sub_match_rec lacc lc =
       match lc with
-      | [] -> next ()
+      | [] -> TailCall next
       | (env, c) :: tl ->
         let mk_ctx ce = mk_ctx (List.rev_append lacc (ce :: List.map snd tl)) in
         let next () = try_sub_match_rec (c :: lacc) tl in
-        aux env c mk_ctx next
+        TailCall (fun () -> aux env c mk_ctx next)
     in
-    try_sub_match_rec [] lc in
-  let lempty () = IStream.Nil in
+    TailCall (fun () -> try_sub_match_rec [] lc) in
+  let lempty () = Fin (IStream.Nil) in
   let result () = aux env c (fun x -> x) lempty in
-  IStream.thunk result
+  IStream.thunk (tlift result)
 
 let match_subterm env sigma pat c = sub_match env sigma pat c
 
