diff --git a/interp/notation.ml b/interp/notation.ml
index 854ae85..26b1e69 100644
--- a/interp/notation.ml
+++ b/interp/notation.ml
@@ -428,8 +428,13 @@ module Numeral = struct
 
 let eval_constr env sigma (c : Constr.t) =
   let c = EConstr.of_constr c in
-  let sigma,t = Typing.type_of env sigma c in
-  let c' = Vnorm.cbv_vm env sigma c t in
+  let c' =
+    if Coq_config.bytecode_compiler then
+      let sigma,t = Typing.type_of env sigma c in
+      Vnorm.cbv_vm env sigma c t   (* vm_compute *)
+    else
+      Tacred.compute env sigma c   (* compute *)
+  in
   EConstr.Unsafe.to_constr c'
 
 (* For testing with "compute" instead of "vm_compute" :
